src/Stratis.Bitcoin/FinalizedBlockInfoRepository.cs

```diff
@@ -0,0 +1,208 @@
+        /// <param name="hash">Block hash.</param>
+        /// <param name="height">Block height.</param>
+        /// <returns><c>true</c> if new value was set, <c>false</c> if <paramref name="height"/> is lower or equal than current value.</returns>
+        bool SaveFinalizedBlockHashAndHeight(uint256 hash, int height);
+    }
+
+    public class FinalizedBlockInfoRepository : IFinalizedBlockInfoRepository
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Access to DBreeze database.</summary>
+        private readonly DBreezeEngine dbreeze;
+
+        /// <summary>Database key under which the block height of the last finalized block height is stored.</summary>
+        private static readonly byte[] finalizedBlockKey = new byte[0];
+
+        /// <summary>Height and hash of a block that can't be reorged away from.</summary>
+        private HashHeightPair finalizedBlockInfo;
+
+        /// <summary>Queue of finalized infos to save.</summary>
+        /// <remarks>All access should be protected by <see cref="queueLock"/></remarks>
+        private readonly Queue<HashHeightPair> finalizedBlockInfosToSave;
+
+        /// <summary>Protects access to <see cref="finalizedBlockInfosToSave"/>.</summary>
+        private readonly object queueLock;
+
+        /// <summary>Task that continously persists finalized block info to the database.</summary>
+        private readonly Task finalizedBlockInfoPersistingTask;
+
+        private readonly CancellationTokenSource cancellation;
+
+        private readonly AsyncManualResetEvent queueUpdatedEvent;
+
+        public FinalizedBlockInfoRepository(string folder, ILoggerFactory loggerFactory)
+        {
+            Guard.NotEmpty(folder, nameof(folder));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            Directory.CreateDirectory(folder);
+            this.dbreeze = new DBreezeEngine(folder);
+
+            this.finalizedBlockInfosToSave = new Queue<HashHeightPair>();
+            this.queueLock = new object();
+
+            this.queueUpdatedEvent = new AsyncManualResetEvent(false);
+            this.cancellation = new CancellationTokenSource();
+            this.finalizedBlockInfoPersistingTask = this.PersistFinalizedBlockInfoContinuouslyAsync();
+        }
+
+        public FinalizedBlockInfoRepository(DataFolder dataFolder, ILoggerFactory loggerFactory)
+            : this(dataFolder.FinalizedBlockInfoPath, loggerFactory)
+        {
+        }
+
+        private async Task PersistFinalizedBlockInfoContinuouslyAsync()
+        {
+            this.logger.LogTrace("()");
+
+            while (!this.cancellation.IsCancellationRequested)
+            {
+                try
+                {
+                    await this.queueUpdatedEvent.WaitAsync(this.cancellation.Token).ConfigureAwait(false);
```

Comment:
Why not AsyncQueue? exactly the same result.

Reply:
We don't know how many blocks are in the queue. It doesn't implement .Count and therefore we can't decide if we got the last item or there are more. 
And we need to save only last item.

Reply:
Ah yes that is a limitation.